Matplotlib - TkAgg backend not available

## Issue description

While going through the [scikit-learn](http://scikit-learn.org/) documentation I came about a [Python script](http://scikit-learn.org/stable/auto_examples/applications/svm_gui.html#example-applications-svm-gui-py) that explicitly requires matplotlib with the TkAgg backend. I tried to generate a nix-shell with matplotlib and tcl/tk and tkinter. However, the script does not execute and instead I get the following error message:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/nix/store/qg48n5n2bc3v83zn2252izwxc1kx2ac4-python-2.7.11-env/lib/python2.7/site-packages/matplotlib/backends/backend_tkagg.py", line 13, in <module>
    import matplotlib.backends.tkagg as tkagg
  File "/nix/store/qg48n5n2bc3v83zn2252izwxc1kx2ac4-python-2.7.11-env/lib/python2.7/site-packages/matplotlib/backends/tkagg.py", line 9, in <module>
    from matplotlib.backends import _tkagg
ImportError: cannot import name _tkagg
```

### Steps to reproduce

In a simplified form:

 * Start nix-shell: 
```
$ nix-shell -p 'with import <nixpkgs> {}; pythonFull.buildEnv.override { extraLibs = with pythonPackages; [ matplotlib tcl tk tkinter ]; }'
```
 * Start Python:
```
$ python
```
 * Try to import matplotlib, configure the backend, and import parts of Tk:
```
>>> import matplotlib as mpl
>>> mpl.use("TkAgg")
>>> from matplotlib.backends.backend_tkagg import FigureCanvasTkAgg
```
 * Observe the following error message:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/nix/store/qg48n5n2bc3v83zn2252izwxc1kx2ac4-python-2.7.11-env/lib/python2.7/site-packages/matplotlib/backends/backend_tkagg.py", line 13, in <module>
    import matplotlib.backends.tkagg as tkagg
  File "/nix/store/qg48n5n2bc3v83zn2252izwxc1kx2ac4-python-2.7.11-env/lib/python2.7/site-packages/matplotlib/backends/tkagg.py", line 9, in <module>
    from matplotlib.backends import _tkagg
ImportError: cannot import name _tkagg
```

## Technical details

* System: NixPkgs on openSuse 13.2 (Harlequin) (x86_64)
* Nix version: nix-env (Nix) 1.10
* Nixpkgs version: "16.09pre83147.df89584"
* Matplotlib Maintainer: @lovek323

I made an attempt to build matplotlib against tcl/tk with the following patch:

```
 pkgs/development/python-modules/matplotlib/default.nix | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/pkgs/development/python-modules/matplotlib/default.nix b/pkgs/development/python-modules/matplotlib/default.nix
index c60ee54..a78b32b 100644
--- a/pkgs/development/python-modules/matplotlib/default.nix
+++ b/pkgs/development/python-modules/matplotlib/default.nix
@@ -4,12 +4,21 @@
 , enableGhostscript ? false, ghostscript ? null, gtk3
 , enableGtk2 ? false, pygtk ? null, gobjectIntrospection
 , enableGtk3 ? false, cairo
+, tcl ? null, tk ? null, tkinter ? null, libX11 ? null
 , Cocoa, Foundation, CoreData, cf-private, libobjc, libcxx
 }:
 
 assert enableGhostscript -> ghostscript != null;
 assert enableGtk2 -> pygtk != null;
 
+let
+  withTk = (tcl != null)
+        && (tk != null)
+        && (tkinter != null)
+        && (libX11 != null)
+         ;
+in
+
 buildPythonPackage rec {
   name = "matplotlib-${version}";
   version = "1.5.1";
@@ -33,7 +42,8 @@ buildPythonPackage rec {
       libpng pkgconfig mock pytz  
     ]
     ++ stdenv.lib.optional enableGtk2 pygtk
-    ++ stdenv.lib.optionals enableGtk3 [ cairo pycairo gtk3 gobjectIntrospection pygobject3 ];
+    ++ stdenv.lib.optionals enableGtk3 [ cairo pycairo gtk3 gobjectIntrospection pygobject3 ]
+    ++ stdenv.lib.optionals withTk [ tcl tk tkinter libX11 ];
 
   patches = stdenv.lib.optionals stdenv.isDarwin [ ./darwin-stdenv.patch ];
```

Unfortunately, compilation fails with the following error message:

```
$ nix-build --pure -A pythonPackages.matplotlib
[...]
gcc -fno-strict-aliasing -g -O2 -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -DPY_ARRAY_UNIQUE_SYMBOL=MPL_matplotlib_backends__tkagg_ARRAY_API -DNPY_NO_DEPRECATED_API=NPY_1_7_API_VERSION -I/nix/store/ag2r3vkbkjxndmv993aliqslc67gj5m0-python2.7-numpy-1.11.0/lib/python2.7/site-packages/numpy/core/include -I/usr/local/include -I/usr/include -I. -I/usr/local/include -I/usr/local/include -Iextern/agg24-svn/include -I/nix/store/adnbhcy51b4x19fmfci6h0qfy5sarwpr-python-2.7.11/include/python2.7 -c src/_tkagg.cpp -o build/temp.linux-x86_64-2.7/src/_tkagg.o
g++ -pthread -shared -lgcc_s build/temp.linux-x86_64-2.7/src/py_converters.o build/temp.linux-x86_64-2.7/src/_tkagg.o -L/usr/local/lib -L/usr/local/lib64 -L/usr/lib -L/usr/lib64 -L/usr/local/lib -L/usr/local/lib -L/nix/store/adnbhcy51b4x19fmfci6h0qfy5sarwpr-python-2.7.11/lib -ltcl -ltk -lpython2.7 -o build/lib.linux-x86_64-2.7/matplotlib/backends/_tkagg.so
/nix/store/s0kpgl8a101lkymj599f5a04s4wflf5s-binutils-2.26-dev/bin/ld: cannot find -ltcl
/nix/store/s0kpgl8a101lkymj599f5a04s4wflf5s-binutils-2.26-dev/bin/ld: cannot find -ltk
collect2: error: ld returned 1 exit status
error: command 'g++' failed with exit status 1
builder for ‘/nix/store/scfk45k4w5qmrd2vccavz6f1p5nb1kkx-python2.7-matplotlib-1.5.1.drv’ failed with exit code 1
error: build of ‘/nix/store/scfk45k4w5qmrd2vccavz6f1p5nb1kkx-python2.7-matplotlib-1.5.1.drv’ failed
```

The linker cannot find the tcl/tk libraries. The linker flags are quite off: They include `-L/usr/local/lib` and the like. This must be due to matplotlib's own build script, as no such flags are set in my bash environment and `nix-build` is executed with the `--pure` flag anyway.

--- a/pkgs/development/python-modules/matplotlib/basedirlist.patch
+++ b/pkgs/development/python-modules/matplotlib/basedirlist.patch
@@ -1,0 +1,8 @@
+diff --git a/setup.cfg b/setup.cfg
+new file mode 100644
+index 0000000..6f81985
+--- /dev/null
++++ b/setup.cfg
+@@ -0,0 +1,2 @@
++[directories]
++basedirlist = .

--- a/pkgs/development/python-modules/matplotlib/default.nix
+++ b/pkgs/development/python-modules/matplotlib/default.nix
@@ -4,11 +4,17 @@
 , enableGhostscript ? false, ghostscript ? null, gtk3
 , enableGtk2 ? false, pygtk ? null, gobjectIntrospection
 , enableGtk3 ? false, cairo
+, enableTk ? false, tcl ? null, tk ? null, tkinter ? null, libX11 ? null
 , Cocoa, Foundation, CoreData, cf-private, libobjc, libcxx
 }:
 
 assert enableGhostscript -> ghostscript != null;
 assert enableGtk2 -> pygtk != null;
+assert enableTk -> (tcl != null)
+                && (tk != null)
+                && (tkinter != null)
+                && (libX11 != null)
+                ;
 
 buildPythonPackage rec {
   name = "matplotlib-${version}";
@@ -33,11 +39,26 @@ buildPythonPackage rec {
       libpng pkgconfig mock pytz  
     ]
     ++ stdenv.lib.optional enableGtk2 pygtk
-    ++ stdenv.lib.optionals enableGtk3 [ cairo pycairo gtk3 gobjectIntrospection pygobject3 ];
+    ++ stdenv.lib.optionals enableGtk3 [ cairo pycairo gtk3 gobjectIntrospection pygobject3 ]
+    ++ stdenv.lib.optionals enableTk [ tcl tk tkinter libX11 ];
 
   patches =
     [ ./basedirlist.patch ] ++
     stdenv.lib.optionals stdenv.isDarwin [ ./darwin-stdenv.patch ];
+
+  # Matplotlib tries to find Tcl/Tk by opening a Tk window and asking the
+  # corresponding interpreter object for its library paths. This fails if
+  # `$DISPLAY` is not set. The fallback option assumes that Tcl/Tk are both
+  # installed under the same path which is not true in Nix.
+  # With the following patch we just hard-code these paths into the install
+  # script.
+  postPatch =
+    let
+      inherit (stdenv.lib.strings) substring;
+      tcl_tk_cache = ''"${tk}/lib", "${tcl}/lib", "${substring 0 3 tk.version}"'';
+    in
+    stdenv.lib.optionalString enableTk
+      "sed -i '/self.tcl_tk_cache = None/s|None|${tcl_tk_cache}|' setupext.py";
 
   checkPhase = ''
     ${python.interpreter} tests.py

--- a/pkgs/development/python-modules/matplotlib/default.nix
+++ b/pkgs/development/python-modules/matplotlib/default.nix
@@ -35,7 +35,9 @@ buildPythonPackage rec {
     ++ stdenv.lib.optional enableGtk2 pygtk
     ++ stdenv.lib.optionals enableGtk3 [ cairo pycairo gtk3 gobjectIntrospection pygobject3 ];
 
-  patches = stdenv.lib.optionals stdenv.isDarwin [ ./darwin-stdenv.patch ];
+  patches =
+    [ ./basedirlist.patch ] ++
+    stdenv.lib.optionals stdenv.isDarwin [ ./darwin-stdenv.patch ];
 
   checkPhase = ''
     ${python.interpreter} tests.py

